crates/client/settlement/src/sync_state.rs

```diff
@@ -0,0 +1,266 @@
+            }
+
+            if sync_from == sync_to {
+                log::info!("[settlement] Verifying state root for block {}", sync_to);
+                Self::verify_starknet_state(substrate_client, &last_settled_state)?;
+                continue;
+            }
+
+            log::info!("[settlement] Syncing state for blocks {} -> {}", sync_from, sync_to);
+            while sync_from < sync_to {
+                let (next_block, substrate_block_hash) = if sync_from + 1 == sync_to {
+                    // This is a typical scenario when we are up to speed with the chain
+                    (block.clone(), notification.hash)
+                } else {
+                    Self::get_starknet_block(substrate_client, sync_from + 1)?
+                };
+
+                let new_state = Self::update_starknet_state(
+                    substrate_client,
+                    settlement_provider,
+                    &last_settled_state,
+                    &next_block,
+                    substrate_block_hash,
+                    starknet_spec,
+                )
+                .await?;
+
+                log::debug!("[settlement] State transitioned to {:?}", new_state);
+                last_settled_state = new_state;
+                sync_from += 1;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Returns Starknet block given it's height (level, number).
+    /// The trick here is that Starknet blocks are embedded into Substrate blocks.
+    ///
+    /// Firstly, we need to get Substrate block hash by the Starknet block height.
+    /// This mapping is kept in a separate storage and there is a dedicated thread which maintains
+    /// it. There might be situations when we cannot resolve the query (e.g. our node is out of
+    /// sync), but eventually it will be ok.
+    ///
+    /// Secondly, we try to find a corresponding Substrate block (header) by its hash.
+    /// Lastly, we extract Starknet block from the Substrate block digest.
+    fn get_starknet_block(substrate_client: &SC, block_number: u64) -> Result<(StarknetBlock, B::Hash), B> {
+        let substrate_block_hash = substrate_client
+            .hash(UniqueSaturatedInto::unique_saturated_into(block_number))?
+            .ok_or_else(|| Error::UnknownStarknetBlock(block_number))?;
+
+        let substrate_block_header = substrate_client
+            .header(substrate_block_hash)?
+            .ok_or_else(|| Error::UnknownSubstrateBlock(substrate_block_hash))?;
+
+        let starknet_block = find_starknet_block(substrate_block_header.digest())?;
+
+        Ok((starknet_block, substrate_block_hash))
+    }
+
+    /// Checks that settlement contract is initialized with the same program & config hash as
+    /// Madara.
+    async fn verify_starknet_spec<SP>(settlement_provider: &SP, starknet_spec: &StarknetSpec) -> Result<(), B>
+    where
+        SP: ?Sized + SettlementProvider<B>,
+    {
+        let spec = settlement_provider.get_chain_spec().await?;
```

Comment:
Is this value going to change once the settlement contract is deployed? If no, we can query it once at the beginning of the thread and use it forever after

Reply:
Maybe it's an assumption we don't want to do?